service/src/DCGMIOGroup.cpp

```diff
@@ -0,0 +1,551 @@
+ *       notice, this list of conditions and the following disclaimer.
+ *
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in
+ *       the documentation and/or other materials provided with the
+ *       distribution.
+ *
+ *     * Neither the name of Intel Corporation nor the names of its
+ *       contributors may be used to endorse or promote products derived
+ *       from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY LOG OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+
+#include <cmath>
+
+#include <fstream>
+#include <iostream>
+#include <sstream>
+#include <string>
+#include <cstring>
+#include <sched.h>
+#include <errno.h>
+
+#include "DCGMIOGroup.hpp"
+#include "DCGMDevicePool.hpp"
+#include "geopm/PlatformTopo.hpp"
+#include "geopm/Exception.hpp"
+#include "geopm/Agg.hpp"
+#include "geopm/Helper.hpp"
+
+namespace geopm
+{
+    DCGMIOGroup::DCGMIOGroup()
+        : DCGMIOGroup(platform_topo(), dcgm_device_pool())
+    {
+    }
+
+    // Set up mapping between signal and control names and corresponding indices
+    DCGMIOGroup::DCGMIOGroup(const PlatformTopo &platform_topo, DCGMDevicePool &device_pool)
+        : m_platform_topo(platform_topo)
+        , m_dcgm_device_pool(device_pool)
+        , m_is_batch_read(false)
+        , m_signal_available({{"DCGM::SM_ACTIVE", {
+                                  "SM activity expressed as a ratio of cycles",
+                                  {},
+                                  [this](unsigned int domain_idx) -> double
+                                  {
+                                      return this->m_dcgm_device_pool.sample(
+                                                   domain_idx, geopm::DCGMDevicePool::M_FIELD_ID_SM_ACTIVE);
+                                  },
+                                  Agg::average,
+                                  string_format_double
+                                  }},
+                              {"DCGM::SM_OCCUPANCY", {
+                                  "Warp residency expressed as a ratio of maximum warps per cycles",
```

Comment:
```suggestion
                                  "Warp residency expressed as a ratio of maximum warps per cycle",
```

One cycle or many cycles?

Reply:
ratio of maximum warps (per cycles is incorrect)